### ISSUE TITLE
Add full support for downloading versions of Minecraft from Omniarchive's collection

### ISSUE THREAD

**Issue by TheMCNerd2017** [PrismLauncher/PrismLauncher#101]
### Goal

Allow players to download versions of Minecraft from Omniarchive's collection within the launcher, as well as fixing bugs/issues in old classic versions caused by new hardware and software.

### Motivation

Omniarchive has archived a lot of versions of Java Minecraft that Mojang/Microsoft do not officially host. Currently to play them through the launcher you have to download the version you want from their collection manually, create a new instance with a similar enough version to the one you want to play, then set the game version to custom and select the JAR file you've downloaded. You can also use Betacraft to play all the versions from Omniarchive's collection, but it requires installing another program on your device and creates possible fragmentation of where and how your Minecraft game data is stored.

If Prism Launcher were to support downloading versions of Minecraft from Omniarchive, more users of Minecraft could easily try out older versions of Minecraft not available from Mojang/Microsoft without having to manually download JAR files or install extra launcher and mods.

### Specification

A decent place to put the functionality would be in both the instance creation menu as well as the version picker that comes up when changing the Minecraft version an instance uses. 

For the instance creation menu, it would be a checkbox on the main Vanilla versions page. The checkbox would have 'Show additional versions from Omniarchive. (?)' with the question mark leading to a wiki page explaining the new functionality. Additional filters would also be added upon ticking the checkbox (Infdev, Indev, Classic and Pre-classic) that would make it easier to find versions from those development points. Additionally, display names of existing versions from Mojang/Microsoft would also be changed to be more accurate when Omniarchive versions are enabled, so users can tell all the versions apart far easier.

For the version picker menu when changing the game version, it would be more simplistic, with a checkbox that toggles the display of Omniarchive versions in the list. If it's preferable to rework the UI of the version picker menu instead, then it could be modified to have a similar layout to the instance creation menu, omitting any features not related to changing the game version (like modpack stuff, instance name, instance group, etc). 

Additionally the metadata server would have to include metadata allowing the launcher to see and download the versions from Omniarchive's collection as well as additional metadata for alternative display names of Mojang/Microsoft versions that have incorrect/inaccurate display names.

### Drawbacks

- ~~Omniarchive's collection is hosted on the Internet Archive, which means that downloads may be slow and files could be unavailable if Internet Archive goes under maintenance. Error handling for Internet Archive errors would have to be added.~~ _No longer necessary; see below._
- ~~Internet Archive restricts downloading of certain files to people that have archive.org accounts if they become too popular, so support for signing in to and using archive.org accounts would need to be added as well.~~ _Omniarchive no longer uses Internet Archive for storing their collection, meaning archive.org account integration and Internet Archive error handling is no longer necessary._
- There is a very slim chance that Microsoft/Mojang files a DMCA claim against the hosted files, resulting in them becoming unavailable to the public. If this were to happen the functionality would be broken until mirrors are found. _While currently the Mojang IP enforcement team are OK with Omniarchive and even present in the community, Microsoft could change their mind about OmniArchive in the future for any reason._

### Unresolved Questions

- ~~Omniarchive has also archived certain versions of the China edition of Minecraft built off of the Java Edition codebase. Should those also be made available for download if support for Omniarchive's collection is added?~~ _This will almost certainly not make it in, since the China edition of Minecraft has lots of spyware embedded within it._ 
- ~~For the tweaks section: Should all of the Betacraft fixes and tweaks be added, or should it be kept only to things launcher-independent (like RetroWrapper and Betacraft proxy)? Additionally are the people behind the tweaks and fixes OK with Prism Launcher being able to download and use them?~~ _Tweaks section would not be required once all fixes listed in [this comment](https://github.com/PrismLauncher/PrismLauncher/issues/101#issuecomment-1339772145) are implemented into the launcher._
- ~~Are the Omniarchive folk OK with allowing their collection to be accessed through Prism Launcher?~~ _The Omniarchive folk are OK with their collection being accessible through the launcher._

### Alternatives Considered

If it's too difficult to implement full Omniarchive support, then maybe another metadata server could be set up which replaces the normal Vanilla version list with versions available from Omniarchive? Though users would have to switch back and forth between the normal metadata server and one meant for the Omniarchive versions.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

The original idea (Omniarchive support) was suggested multiple times in the old repository alongside someone else suggesting Betacraft support, so I decided to suggest it here since I thing having full Omniarchive support would be great for the launcher.

Additionally the end result of the suggestion is not meant to replace or harm Betacraft in any way, it's only meant to make it easier to access old versions of Minecraft without having to manually download JAR files or install extra launchers and mods.

**Comment by ghost** [comment]
Actually, how does one import a .jar into Prism?

**Comment by TheMCNerd2017** [comment]
> Actually, how does one import a .jar into Prism?

I think you click on the 'Replace minecraft.jar' button. You can then choose a JAR file of the game. You also have to make sure to choose a version of the game (when initially creating the instance) that is closest to the JAR file you're going to use.

**Comment by Scrumplex** [comment]
A lot of them were added in PrismLauncher/meta#5 already

**Comment by TheKodeToad** [comment]
Some of the suggestions here seem to relate to the actual launcher, but maybe it would make more sense being moved to the meta repo?

Some of the legacy version fixes could be implemented in a similar way to #443 - although more intrusive patches would need to be implemented in an agent which is probably out of scope for NewLaunch (the Java component of Prism).

**Comment by forkiesassds** [comment]
A good idea for one of the client fixes is to un-applet the whole game to fix issues with HiDPI and M1 iirc.

**Comment by TheKodeToad** [comment]
> A good idea for one of the client fixes is to un-applet the whole game to fix issues with HiDPI and M1 iirc.

This can already be done with the noapplet trait. This does, however, ignore the window dimension settings. It also doesn't work on all versions.

**Comment by forkiesassds** [comment]
> > A good idea for one of the client fixes is to un-applet the whole game to fix issues with HiDPI and M1 iirc.
> 
> This can already be done with the noapplet trait. This does, however, ignore the window dimension settings. It also doesn't work on all versions.

Which versions does it not work on? Also by that I actually meant making the game not use AWT and Canvas...

**Comment by TheKodeToad** [comment]
> Which versions does it not work on? Also by that I actually meant making the game not use AWT and Canvas...

On some versions such as classic and infdev there is no main class. What are the specific issues on M1 macs, btw? I'm interested.

**Comment by forkiesassds** [comment]
> > Which versions does it not work on? Also by that I actually meant making the game not use AWT and Canvas...
> 
> On some versions such as classic and infdev there is no main class. What are the specific issues on M1 macs, btw? I'm interested.

IIRC the main issue on M1 is the inverted color issue which is caused by Canvas and AWT.

**Comment by MisterSheeple** [comment]
Hi there, I'm one of the admins of Omniarchive.

We'd love to see our collection integrated into Prism. Not a lot of launchers support legacy MC versions, and we proposed a similar feature for MultiMC close to 5 years ago, so it'd be nice to see things come full circle.

**Comment by MisterSheeple** [comment]
A few things I'll add to this:

- Archive.org no longer houses our main version repository. We haven't had access to the archive.org item pages for over a year, so they're missing some things. We've since switched to using our own bucket.
- Instead of the Vanilla and Omniarchive buttons in the version picker, I'd like to suggest a more intuitive option like a checkbox that shows Omniarchive versions alongside other versions in the list when toggled.
- It's true that we archive Minecraft China, but we do not suggest any use of it by anyone, as it's got spyware galore.
- In the 5 years that we've been operating, we have never received a DMCA takedown from Mojang for the Java versions. Mojang's IP enforcement team is even present in our Discord, and they've never given us any trouble.

Feel free to let me know if any of you have questions about anything and I'll be happy to help!

**Comment by MisterSheeple** [comment]
Once our collection of binaries is made available, here's a list of things that will need to be fixed for legacy versions to be fully operational:

- [ ] Online mode (alpha/beta auth)
- [ ] Verify names (classic auth)
- [x] Fix skins (#443)
- [ ] Fix sounds (not an easy task because old sounds aren't stored on Mojang's servers anymore)
- [ ] Fix inverted colors on ARM MacOS (#698)
- [ ] Fix crashing on Java 8
- [ ] Implement antipiracy workaround for indev/infdev


Only those last two are "required" for things to work at all, so consider most of these low priority for the time being.

**Comment by MisterSheeple** [comment]
I've created a bit of a concept sketch to show what I think Omniarchive integration could look like during instance creation. Feedback would be greatly appreciated.
![HHTa5Ev](https://github.com/user-attachments/assets/b3f630c0-bdac-4fdf-ba08-0f324781b86b)

**Comment by jopejoe1** [comment]
Wrong data of versions could be corrected here https://github.com/PrismLauncher/meta/blob/master/static/mojang/minecraft-legacy-override.json and the versions from Omniarchive not provided by Mojang could be similar added like I did for the old_snapshots here https://github.com/PrismLauncher/meta/blob/master/static/mojang/minecraft-old-snapshots.json in https://github.com/PrismLauncher/meta/pull/5

**Comment by MisterSheeple** [comment]
> Wrong data of versions could be corrected here https://github.com/PrismLauncher/meta/blob/master/static/mojang/minecraft-legacy-override.json

The incorrect data is a Mojang issue that's gone unfixed for close to a decade. Would it be okay to replace that with our own sources, even when toggled off? I think it might raise some questions.

> and the versions from Omniarchive not provided by Mojang could be similar added like I did for the old_snapshots here https://github.com/PrismLauncher/meta/blob/master/static/mojang/minecraft-old-snapshots.json

Sounds good to me

**Comment by jopejoe1** [comment]
> The incorrect data is a Mojang issue that's gone unfixed for close to a decade. Would it be okay to replace that with our own sources, even when toggled off? I think it might raise some questions.


I don't see a reason as to why we should not apply data that is known to be more accurate, but would need to wait and see what the maintainers are thinking about that.

**Comment by MisterSheeple** [comment]
It seems like it might require changing data anyway, because the more I look at it, the more it seems we'd have to do it across the board to make it more accurate. For instance, some versions, despite being different, share the same version number visually. The way we typically differentiate this in version names is with shortened timestamps such as `a1.0.13_01-1038` and `a1.0.13_01-1444`. And for versions where the original version differs from the launcher version, we append `-launcher` to the end of the launcher version. For instance, `c0.0.13a_03` and `c0.0.13a_03-launcher`. If you'd like to see more examples of this naming scheme, you can refer to our version index [here](https://docs.google.com/spreadsheets/d/1OCxMNQLeZJi4BlKKwHx2OlzktKiLEwFXnmCrSdAFwYQ/view#gid=872531987).

**Comment by MisterSheeple** [comment]
If everyone's okay with the naming convention I've mentioned in the previous comment, then I'll push a PR to the meta repo later today to get further feedback.

**Comment by TheMCNerd2017** [comment]
> * Archive.org no longer houses our main version repository. We haven't had access to the archive.org item pages for over a year, so they're missing some things. We've since switched to using our own bucket.

I didn't know that the main version repository is no longer on archive.org (the main site still points there). I'm guessing that the site is going to be updated at some point to point to the new location?
> * Instead of the Vanilla and Omniarchive buttons in the version picker, I'd like to suggest a more intuitive option like a checkbox that shows Omniarchive versions alongside other versions in the list when toggled.

I agree that a checkbox would be better than two buttons. The reason I thought of two buttons was because the version picker for already existing instances is more limited (does not have filter functions) than the one for creating an instance.
> * It's true that we archive Minecraft China, but we do not suggest any use of it by anyone, as it's got spyware galore.

I'm not surprised that it has a bunch of spyware, and it's OK if it's never available from within a launcher for that reason. I actually was unsure about even including in the RFC hence why I included it in the unresolved questions section.
> * In the 5 years that we've been operating, we have never received a DMCA takedown from Mojang for the Java versions. Mojang's IP enforcement team is even present in our Discord, and they've never given us any trouble.

That's good that Mojang's IP enforcement team has not given you any trouble, however Microsoft could see the project differently in the future depending on various circumstances.
> Feel free to let me know if any of you have questions about anything and I'll be happy to help!

Should I add the updated information to the RFC or should I leave it alone? Also thanks for commenting on the RFC by the way.

**Comment by MisterSheeple** [comment]
>I'm guessing that the site is going to be updated at some point to point to the new location?

Yes, very soon.

>Should I add the updated information to the RFC or should I leave it alone?

Yeah, feel free. Thanks!

**Comment by MisterSheeple** [comment]
Some concerns were brought up in the Prism Launcher Discord server regarding changing the names of existing versions, so how about this:

What if we add a new tag to metadata files that modifies the display names of existing versions *only* when Omniarchive is toggled on? Without having these display names, it gets confusing to tell these versions apart from the ones we have, so I think this is a fair middle ground.

**Comment by TheMCNerd2017** [comment]
> Some concerns were brought up in the Prism Launcher Discord server regarding changing the names of existing versions, so how about this:
> 
> What if we add a new tag to metadata files that modifies the display names of existing versions _only_ when Omniarchive is toggled on? Without having these display names, it gets confusing to tell these versions apart from the ones we have, so I think this is a fair middle ground.

Hello, and sorry for the very late response (my PC had a power supply issue which I finally got fixed). I went and updated the RFC with the new information.

As for the new tag to metadata files that only takes affect when Omniarchive is enabled, that is a good and fair middle ground that allows users to still see the version names that Microsoft/Mojang provides while also seeing the accurate version names.

**Comment by MisterSheeple** [comment]
Hi, bit of an update:

We're working on some proposed changes to the version manifest that should allow for Omniarchive versions to be implemented. As mentioned in one of my previous comments, some changes to Prism's UI will be required in order for implementation to be fully realized. We'll have a demonstration to share here soon for approval on our proposed changes to the manifest, but in short, we plan on doing the following:

- Omni versions will be stored in the same json as versions from Mojang, differentiated with a key stating that the version is from Omni
- Version types will be overhauled to use more specific names instead of lumping everything before release into `old_alpha` and `old_beta`
    - The current types can continue to be used until the needed changes are implemented in Prism's UI itself
- We will be overriding some Mojang metadata, such as:
    - Version names (to make version names more accurate, fix duplicate names and wrong names, etc.)
    - Version types will be modified to use the aforementioned more specific nomenclature 

Subject to change based on any feedback you all may have.

**Comment by Kationor** [comment]
I'm not sure whether *specific* Omniarchive support is the best solution in this case.

On the meta side, full Omniarchive support can be implemented by more general changes to how prism handles custom meta servers, such as:
- allowing users to use multiple meta servers at once (or allow only one at a time but implement quick-switching between them)
- allowing meta servers to specify custom release types (currently those filters are hard-coded in the launcher)
- properly specifying the meta server/manifest structure

This would have a few advantages:
- prism does not need to maintain yet another custom integration
- other projects can also provide their own meta servers easily
- prism avoids potential legal issues if Mojang ever change their mind about tolerating Omniarchive

This also has the significant drawback of reducing Omniarchive's visibility compared to direct integration with the launcher, though Prism could simply have Omni's meta server configured OOTB if we really want to endorse it.

**Comment by MisterSheeple** [comment]
> I'm not sure whether _specific_ Omniarchive support is the best solution in this case.

While I agree that better support for custom meta servers should be a thing, I also think tight integration of the Omni meta server (via a toggle in the instance creation UI, available as a sort of "advanced mode") would be good for users.

> * allowing users to use multiple meta servers at once (or allow only one at a time but implement quick-switching between them)

The way we decided to handle this (per conversations in the Prism discord) would be to have the Omni manifest comprise of both Mojang versions and Omni versions, rather than merging it with the Mojang manifest. We opted to do this because for us, we have versions that conflict with the names of versions in the launcher. For example, multiple versions of 1.16 exist, but Mojang only has one of them in the launcher. So for Omni, we differentiate this by calling these versions 1.16-221349 and 1.16-231620, which are named based on the compile timestamps. If we were to merge our manifest with the Mojang manifest in some way, we would need to provide a lot of overrides for the Mojang manifest to change the name of a version when it switches to Omni mode. This is clearly not ideal, so we opted to just use a separate manifest with Mojang versions included in it.

> * allowing meta servers to specify custom release types (currently those filters are hard-coded in the launcher)

We're developing an extended manifest spec, based on Mojang's `version_manifest_v2` because the existing formats (both Prism's and Mojang's) are unable to fully accommodate for what we're trying to do. We discussed this at length in the Prism discord with members of the Prism development team and decided that this is the way we should handle this. We're adding additional release types to the spec, so it's basically up to the Prism team whether they want the types we're implementing to be hard-coded into the launcher or for it to dynamically change based on what's in the manifest.

If there's anything else you think should be added to our extended manifest spec to benefit other meta servers, let us know here or in the Prism discord.

**Comment by Ryex** [comment]
I'd like to that that *ideally* Prism retains tight control of the metadata structure it accepts. 
the *existance* of  the metadata server both enables Prism to function and presents a unique security risk. your trusting a report serve to tell you what binary blobs to trust and execute. so long as a trusted party is generating that metadata and it has a clear fixed structure we can be reasonably safe.

But if we allow third-party structures to sneak in to the metadata when the launcher isn't prepared to handle all the potential edge cases in data structure problems can crop up.

All this to say that "allowing meta servers to specify custom" *anything* isn't going to happen.


Allowing the user to set a custom meta api url is more an anti dead-ware feature in that the software will still be functional should we for some reason have to shut down our meta server because you can always host your own.

We have some internal proposals to write a full v2 metadata spec

**Comment by Kationor** [comment]
> @Ryex: I'd like to that that *ideally* Prism retains tight control of the metadata structure it accepts.

That is reasonable, but my proposal does not change the structure at all.

> @Ryex: But if we allow third-party structures to sneak in to the metadata when the launcher isn't prepared to handle all the potential edge cases in data structure problems can crop up.

That is a real concern, and edge cases *do* exist even in the current meta format. (ever tried naming a version ""?) That's why this is a capability that should be introduced in a *new* meta version like the one already being discussed.

>@Ryex: All this to say that "allowing meta servers to specify custom" anything isn't going to happen.

The functionality is *already* there. The only thing that prevents meta servers from actually using custom types is the version filter in the UI (see 917f148fc41db7d68ef94c8ea29cd417a8f1ff33). This is simply about removing that artificial limitation.

If you're really *that* concerned about UI edge cases, you don't even need to make the UI dynamic. Just add a "Show all" or "Show others" checkbox to the filter and you're done.

> @Ryex: Allowing the user to set a custom meta api url is more an anti dead-ware feature in that the software will still be functional should we for some reason have to shut down our meta server because you can always host your own.

The setting in the launcher itself doesn't advertise itself that way, and there's no reason the feature needs to be limited to that. Sure, meta server stuff should not be a high priority - the vast majority will be fine with what currently exists - but why limit it when it could be so much more?

> @Ryex: We have some internal proposals to write a full v2 metadata spec

Where do I need to look to catch the first drafts when they are released?

**Comment by Ryex** [comment]
> The functionality is already there. The only thing that prevents meta servers from actually using custom types is the version filter in the UI (see 917f148). This is simply about removing that artificial limitation.

that's more a result of an extremely poor json parseing system we've inherented. it's something we are looking into addressing in the near future as it will be important to get a property type safe structural system in place for reworking our data formats.

all meta work will most likely be taking place over on the @PrismLauncher/meta repo

**Comment by catfromplan9** [comment]
Any updates on this?

**Comment by DexrnZacAttack** [comment]
Unsure if this is useful here, but https://github.com/Omniarchive/manifest exists now.
I wonder how hard it would be to draft up a fork with multi meta server support 🤔

**Comment by catfromplan9** [comment]
Could anyone involved please make it clear to me what the blocker is here or what we've been waiting on? I'm even willing to do a PR myself for whatever needs to happen to get this moving, 4 years on this open issue is getting a bit ridiculous

**Comment by DexrnZacAttack** [comment]
> Could anyone involved please make it clear to me what the blocker is here or what we've been waiting on? I'm even willing to do a PR myself for whatever needs to happen to get this moving, 4 years on this open issue is getting a bit ridiculous

I was also up for it until I got very busy...
Hopefully the impl can be abstracted properly to allow the user to add any arbitrary meta server assuming the API and format is supported (many replicate Mojang's meta format)

and then in code people can add special handling themselves if needed

**Comment by DexrnZacAttack** [comment]
The meta server is currently down for now, while we work on making a replacement.
Will update here when that's finished, and then work on the abstractions.
